cmd/cluster-replication.go

```diff
@@ -0,0 +1,1657 @@
+			return fmt.Errorf("bucket replication config validation error: %#v", apiErr)
+		}
+		err = newReplicationConfig.Validate(bucket, sameTarget)
+		if err != nil {
+			return err
+		}
+		// Config looks good, so we save it.
+		replCfgData, err := xml.Marshal(newReplicationConfig)
+		if err != nil {
+			return err
+		}
+		err = globalBucketMetadataSys.Update(bucket, bucketReplicationConfig, replCfgData)
+		if err != nil {
+			fmt.Printf("%s->%s: replconferr: %v\n", c.state.Name, peer.Name, err)
+		}
+		return err
+	}
+
+	errMap := make(map[string]error, len(c.state.Peers))
+	for d, peer := range c.state.Peers {
+		if d == globalDeploymentID {
+			continue
+		}
+		if err := configurePeerFn(d, peer); err != nil {
+			errMap[d] = err
+		}
+	}
+	return c.toErrorFromErrMap(errMap)
+}
+
+// PeerBucketDeleteHandler - deletes bucket on local in response to a delete
+// bucket request from a peer.
+func (c *ClusterReplMgr) PeerBucketDeleteHandler(ctx context.Context, bucket string, forceDelete bool) error {
+	c.RLock()
+	defer c.RUnlock()
+	if !c.enabled {
+		return errCRNotEnabled
+	}
+
+	// FIXME: need to handle cases where globalDNSConfig is set.
+
+	objAPI := newObjectLayerFn()
+	if objAPI == nil {
+		return errServerNotInitialized
+	}
+	err := objAPI.DeleteBucket(ctx, bucket, forceDelete)
+	if err != nil {
+		return err
+	}
+
+	globalNotificationSys.DeleteBucketMetadata(ctx, bucket)
+
+	return nil
+}
+
+// IAMChangeHook - called when IAM items need to be replicated to peer clusters.
+// This includes named policy creation, policy mapping changes and service
+// account changes.
+//
+// All policies are replicated.
+//
+// Policy mappings are only replicated when they are for LDAP users or groups
+// (as an external IDP is always assumed when CR is used). In the case of
+// OpenID, such mappings are provided from the IDP directly and so are not
+// applicable here.
+//
+// Only certain service accounts can be replicated:
+//
+// Service accounts created for STS credentials using an external IDP: such STS
+// credentials would be valid on the peer clusters as they are assumed to be
+// using the same external IDP. Service accounts when using internal IDP or for
+// root user will not be replicated.
+//
+// STS accounts are replicated, but only if the session token is verifiable
+// using the local cluster's root credential.
+func (c *ClusterReplMgr) IAMChangeHook(ctx context.Context, item madmin.CRIAMItem) error {
+	// The IAM item has already been applied to the local cluster at this
+	// point, and only needs to be updated on all remote peer clusters.
+
+	c.RLock()
+	defer c.RUnlock()
+	if !c.enabled {
+		return nil
+	}
+
+	cErr := c.concDo(nil, func(d string, p madmin.PeerInfo) error {
+		admClient, err := c.getAdminClient(ctx, d)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+
+		err = admClient.CRInternalReplicateIAMItem(ctx, item)
+		if err != nil {
+			fmt.Printf("%s->%s: CRInternalReplicateIAMItem: %v\n", c.state.Name, p.Name, err)
```

Comment:
```suggestion
			fmt.Printf("%s->%s: CRInternalReplicateIAMItem: %v\n", c.state.Name, p.Name, err)
```

Needs to be removed or add `logger.LogIf()` instead